### PR TITLE
Replace the custom post-processing shader tutorial

### PR DESCRIPTION
**Pull request by RolandMarchand**
The original method of applying shaders to viewports did not work in practice. It was also too complicated. This new way of doing it with CanvasLayers works well and is easy to implement.

* I reorganised the information to improve the information flow and readability, and to reduce redundancy.
* I replaced the original cube example with a sheep sprite. While the original tutorial claimed to be in 2D, the example cubes were in 3D, which was confusing. Also, sheep are more engaging than cubes.
* I replaced the the Sobel filter example with an hex pixelization example, because in practice applying a Sobel filter to a frame often makes it appear pure black. People trying out this filter may think it does not work. Removing the Sobel filter example also removes a note about the filter's implementation, saving space. In contrast, a pixelization filter works on nearly all images and makes for a more engaging and clear example.

The total amount of words is around 300.

### CODE REVIEW

**Review by clayjohn — COMMENTED** [review]
Some General notes:

1. Images should be run through an image compression tool (see [here](https://docs.godotengine.org/en/latest/community/contributing/docs_writing_guidelines.html#image-contribution-guidelines) for recommendations)
2. I wonder if it is worthwhile to make this a little simpler by using BackBufferCopy instead. In theory, using BackBufferCopy instead of CanvasLayer should be quite a bit faster and should achieve the same result. I guess the only thing it doesn't do is allow adjusting the canvas layer to manually determine draw order (without moving nodes in the scene tree)

**Comment by RolandMarchand** [comment]
>     1. Images should be run through an image compression tool (see [here](https://docs.godotengine.org/en/latest/community/contributing/docs_writing_guidelines.html#image-contribution-guidelines) for recommendations)
The images have already been compressed using Oxipng.

**Comment by RolandMarchand** [comment]
> I wonder if it is worthwhile to make this a little simpler by using BackBufferCopy instead. In theory, using BackBufferCopy instead of CanvasLayer should be quite a bit faster and should achieve the same result. I guess the only thing it doesn't do is allow adjusting the canvas layer to manually determine draw order (without moving nodes in the scene tree)

I prefer the proposed method because Godot users are familiar with `CanvasLayer` and `ColorRect` nodes, unlike with the `BackBufferCopy` node. Also, the consistency between the single and multi-pass post-processing methods reduces the cognitive load.

File: tutorials/shaders/custom_postprocessing.rst [hunk 1]
```diff
@@ -6,167 +6,114 @@ Custom post-processing
...
-    ``ViewportContainer`` to render 3D objects in a 2D scene.
-
-For this demo, we will use a :ref:`Node2D <class_Node2D>` with a ``ViewportContainer`` and finally a
-``Viewport``. Your **Scene** tab should look like this:
-
-.. image:: img/post_hierarchy1.png
-
-Inside the ``Viewport``, you can have whatever you want. This will contain
-your main scene. For this tutorial, we will use a field of random boxes:
+Post-processing effects are shaders applied to a frame after Godot rendered it. To apply a shader to a frame, create a :ref:`CanvasLayer <class_CanvasLayer>`, and give it a :ref:`ColorRect <class_ColorRect>`. Give it a new :ref:`ShaderMaterial <class_ShaderMaterial>`, and set the ``ColorRect``'s layout to "Full Rect".
```
Comment (aaronfranke) [inline tutorials/shaders/custom_postprocessing.rst:18]: In the RST docs, single newline characters are ignored, so you can split up paragraphs into multiple lines. Note that this is different from the class reference which does care about single newline characters.

File: tutorials/shaders/custom_postprocessing.rst [hunk 2]
```diff
@@ -6,167 +6,114 @@ Custom post-processing
...
 
-    .. code-block:: glsl
+For this demo, we will use this :ref:`Sprite <class_Sprite2D>` of a sheep.
 
-      // Inside the Shader.
-      uniform sampler2D ViewportTexture;
+.. image:: img/post_example1.png
 
-    And you can pass the texture into the shader from GDScript like so:
+To the ``ColorRect``'s ``ShaderMaterial``, assign a new :ref:`Shader <class_Shader>`. You can access the frame's texture and UV with the built in ``SCREEN_TEXTURE`` and ``SCREEN_UV`` uniforms.
```
Comment (aaronfranke) [inline tutorials/shaders/custom_postprocessing.rst:34]:
```suggestion
Assign a new :ref:`Shader <class_Shader>` to the ``ColorRect``'s ``ShaderMaterial``. You can access the frame's texture and UV with the built in ``SCREEN_TEXTURE`` and ``SCREEN_UV`` uniforms.
```

File: tutorials/shaders/custom_postprocessing.rst [hunk 3]
```diff
@@ -6,167 +6,114 @@ Custom post-processing
...
-    .. image:: img/post_outline.png
+.. image:: img/post_example2.png
 
 Multi-pass post-processing
 --------------------------
 
-Some post-processing effects like blur are resource intensive. If you break them down in multiple passes
-however, you can make them run a lot faster. In a multipass material, each pass takes the result from the
-previous pass as an input and processes it.
+Some post-processing effects like blurs are resource intensive. But you can make them run a lot faster if you break them down in multiple passes. In a multipass material, each pass takes the result from the previous pass as an input and processes it.
```
Comment (aaronfranke) [inline tutorials/shaders/custom_postprocessing.rst:71]:
```suggestion
Some post-processing effects like blurs are resource intensive. You can make them run a lot faster if you break them down in multiple passes. In a multipass material, each pass takes the result from the previous pass as an input and processes it.
```

File: tutorials/shaders/custom_postprocessing.rst [hunk 4]
```diff
@@ -6,167 +6,114 @@ Custom post-processing
...
 
-    You can choose not to use a ``ViewportContainer``, but if you do so, you will
-    need to create your own uniform in the shader and pass the ``Viewport`` texture in
-    manually, like so:
+    As of the time of writing, Godot does not support rendering to multiple buffers at the same time. Your
+    post-processing shader will not have access to normals or other render passes. You only have
+    access to the rendered frame.
 
-    .. code-block:: glsl
+For this demo, we will use this :ref:`Sprite <class_Sprite2D>` of a sheep.
```
Comment (aaronfranke) [inline tutorials/shaders/custom_postprocessing.rst:30]:
```suggestion
We will use this image of a sheep as an example.
```

File: tutorials/shaders/custom_postprocessing.rst [hunk 5]
```diff
@@ -6,167 +6,114 @@ Custom post-processing
 Introduction
 ------------
 
-Godot provides many post-processing effects out of the box, including Bloom, DOF, and SSAO. Sometimes you
-want to write your own custom effect. Here's how you can do so.
-
-Post-processing effects are shaders applied to a frame after Godot rendered it. You first want to render
-your scene into a :ref:`Viewport <class_Viewport>`, then render the ``Viewport``
-inside a :ref:`ViewportTexture <class_ViewportTexture>` and show it on the screen.
+Godot provides many post-processing effects out of the box, including Bloom, DOF, and SSAO. Sometimes you want to write your own custom effect. Here's how you can do so.
```
Comment (aaronfranke) [inline tutorials/shaders/custom_postprocessing.rst:9]:
```suggestion
Godot provides many post-processing effects out of the box, including Bloom, DOF, and SSAO. However, advanced use cases may require custom effects. This article explains how to write your own custom effects.
```

**Comment by RolandMarchand** [comment]
I have received explicit permission from Arlex80 about using his hex pixelizer.
image

**Comment by mhilbrunner** [comment]
> > Images should be run through an image compression tool (see [here](https://docs.godotengine.org/en/latest/community/contributing/docs_writing_guidelines.html#image-contribution-guidelines) for recommendations)
> 
> The images have already been compressed using Oxipng.

Does this apply to `[tutorials/shaders/img/post_example1.png](https://github.com/godotengine/godot-docs/pull/5946/files#diff-53f10ff2ff29b60e9c7e8df56ab3887f8112b9e7ab550073ac9d414dfa18f439)` as well? ~1mb is a bit on the large-ish site for a an image in a docs page. Maybe we can compress it a bit further or reduce the resolution?

Besides that, I spotted nothing else and this looks good, docs guidelines wise. @clayjohn Okay if I leave this in your capable hands? :)

**Comment by RolandMarchand** [comment]
> > > Images should be run through an image compression tool (see [here](https://docs.godotengine.org/en/latest/community/contributing/docs_writing_guidelines.html#image-contribution-guidelines) for recommendations)
> > 
> > 
> > The images have already been compressed using Oxipng.
> 
> Does this apply to `tutorials/shaders/img/post_example1.png` as well? ~1mb is a bit on the large-ish site for a an image in a docs page. Maybe we can compress it a bit further or reduce the resolution?
> 
> Besides that, I spotted nothing else and this looks good, docs guidelines wise. @clayjohn Okay if I leave this in your capable hands? :)

I can try to compress it even further.

**Review by mhilbrunner — APPROVED** [review]
Thanks, I think 600kb is okay now.

**Comment by RolandMarchand** [comment]
In the previous commit, I have added a note about `BackBufferCopy`, I have mentioned the shader application precedence based on the `CanvasLayer` order, and I have filled the paragraphs to 80 characters.

**Review by clayjohn — APPROVED** [review]

**Comment by mhilbrunner** [comment]
Merged! What an amazing first contribution to Godot's docs :) Congrats and thank you! 🎉 

Also thank you to all the reviewers!

**Comment by Calinou** [comment]
The cherry-pick doesn't apply cleanly, and there are probably some renames required to match the 3.x API. Could you open a separate pull request targeting the `3.5` branch? Thanks in advance :slightly_smiling_face:

**Pull request merged**